File: Java/Graphs/ProblemsOnDfsBfs/547-number-of-provinces/number-of-provinces.java
class Solution {
    private void dfs(int node, ArrayList<ArrayList<Integer>> adjList, int visited[]){
        visited[node] = 1;
        for(Integer it:adjList.get(node)){
            if(visited[it] == 0){
                dfs(it,adjList,visited);
            }
        }
    }
    public int findCircleNum(int[][] isConnected) {
        ArrayList<ArrayList<Integer>> adjList = new ArrayList<>();

        for(int i=0;i<isConnected.length;i++){
            adjList.add(new ArrayList<>());
        }

        //convert adj matrix to list
        for(int i=0;i<isConnected.length;i++){
            for(int j =0;j<isConnected[0].length;j++){
                if(isConnected[i][j] == 1 && i!=j){
                    adjList.get(i).add(j);
                    adjList.get(j).add(i);
                }
            }
        }
        int visited[] = new int[isConnected.length];
        int cnt=0;
        for(int i=0;i<visited.length;i++){
            if(visited[i] ==0){
                cnt++;
                dfs(i,adjList,visited);
            }
        }
        return cnt;
    }
}

File: Python/Graphs/ProblemsOnBFSAndDFS/547-number-of-provinces/number-of-provinces.py
class Solution:
    
    def bfs(self, start : int, visited : List[int], grid: List[List[int]]) -> None:
        visited[start] = 1

        q = deque()

        q.append(start)

        n = len(grid)

        while q:
            node = q.popleft()
            for neighbour in range(n):
                if grid[node][neighbour] == 1 and visited[neighbour] == 0:
                    visited[neighbour] = 1
                    q.append(neighbour)

    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)

        visited = [0]*n

        cnt = 0

        for i in range(n):
                if visited[i] == 0:
                    cnt+=1
                    self.bfs(i,visited,isConnected)
                
        return cnt
        